### PR TITLE
fix(sql): fix race condition in `where` execution logic

### LINKED ISSUE #2454: Regex filter may throw StringIndexOutOfBoundsException
### Describe the bug

Originates from #2441

### To reproduce

Suspicious offending query is:
```sql
SELECT
  timestamp as time,
  exchange,
  symbol,
  count() as trades
FROM
  table
WHERE
  symbol ~ '.*-?ETH' and
  timestamp BETWEEN '2022-08-17T13:05:43.981Z' AND '2022-08-17T19:05:43.981Z'
SAMPLE BY 1m
```
results in
```
2022-08-17T19:05:49.805401Z E i.q.c.p.PGWireServer
java.lang.StringIndexOutOfBoundsException: String index out of range: 8
        at java.lang.StringLatin1.charAt(StringLatin1.java:48)
        at java.lang.String.charAt(String.java:1515)
        at java.lang.Character.codePointAt(Character.java:8910)
        at java.util.regex.Pattern$CharPropertyGreedy.match(Pattern.java:4273)
        at java.util.regex.Pattern$Start.match(Pattern.java:3608)
        at java.util.regex.Matcher.search(Matcher.java:1728)
        at java.util.regex.Matcher.find(Matcher.java:745)
        at io.questdb.griffin.engine.functions.regex.MatchStrFunctionFactory$MatchConstPatternFunction.getBool(MatchStrFunctionFactory.java:100)
        at io.questdb.griffin.engine.table.AsyncFilteredRecordCursorFactory.filter(AsyncFilteredRecordCursorFactory.java:167)
        at io.questdb.cairo.sql.async.PageFrameReduceJob.reduce(PageFrameReduceJob.java:175)
        at io.questdb.cairo.sql.async.PageFrameReduceJob.consumeQueue(PageFrameReduceJob.java:132)
        at io.questdb.cairo.sql.async.PageFrameReduceJob.consumeQueue(PageFrameReduceJob.java:106)
        at io.questdb.cairo.sql.async.PageFrameSequence.stealWork(PageFrameSequence.java:390)
        at io.questdb.cairo.sql.async.PageFrameSequence.dispatch(PageFrameSequence.java:369)
        at io.questdb.cairo.sql.async.PageFrameSequence.next(PageFrameSequence.java:294)
        at io.questdb.griffin.engine.table.AsyncFilteredRecordCursor.fetchNextFrame(AsyncFilteredRecordCursor.java:184)
        at io.questdb.griffin.engine.table.AsyncFilteredRecordCursor.of(AsyncFilteredRecordCursor.java:226)
        at io.questdb.griffin.engine.table.AsyncFilteredRecordCursorFactory.getCursor(AsyncFilteredRecordCursorFactory.java:129)
        at io.questdb.griffin.engine.table.SelectedRecordCursorFactory.getCursor(SelectedRecordCursorFactory.java:58)
        at io.questdb.griffin.engine.groupby.SampleByFillNoneRecordCursorFactory.getCursor(SampleByFillNoneRecordCursorFactory.java:97)
        at io.questdb.cutlass.pgwire.PGConnectionContext.setupFactoryAndCursor(PGConnectionContext.java:2488)
        at io.questdb.cutlass.pgwire.PGConnectionContext$PGConnectionBatchCallback.postCompile(PGConnectionContext.java:2593)
        at io.questdb.griffin.SqlCompiler.compileBatch(SqlCompiler.java:930)
        at io.questdb.cutlass.pgwire.PGConnectionContext.processQuery(PGConnectionContext.java:2256)
        at io.questdb.cutlass.pgwire.PGConnectionContext.parse(PGConnectionContext.java:1550)
        at io.questdb.cutlass.pgwire.PGConnectionContext.handleClientOperation(PGConnectionContext.java:415)
        at io.questdb.cutlass.pgwire.PGJobContext.handleClientOperation(PGJobContext.java:81)
        at io.questdb.cutlass.pgwire.PGWireServer$1.lambda$$0(PGWireServer.java:81)
        at io.questdb.network.AbstractIODispatcher.processIOQueue(AbstractIODispatcher.java:166)
        at io.questdb.cutlass.pgwire.PGWireServer$1.run(PGWireServer.java:106)
        at io.questdb.mp.Worker.run(Worker.java:116)
```

Reported by user on the built-in JDK (QuestDB rt distribution). The distinct values of symbol column were the following:
```
      "ALT-USD-22U30"
      "BTC-USD"
      "BTC-USD-22U30"
      "BTC-USD-PERP"
      "BTC-USDT"
      "CEL-USD"
      "CEL-USD-22U30"
      "CEL-USD-PERP"
      "ETH-USD"
      "ETH-USD-PERP"
      "ETH-USDT"
      "EXCH-USD-22U30"
      "MID-USD-22U30"
      "PRIV-USD-22U30"
      "SHIT-USD-22U30"
      "SOL-USD"
      "SOL-USD-22U30"
      "SOL-USD-PERP"
      "SOL-USDT"
      "USDT-USD"
      "USDT-USD-22U30"
      "USDT-USD-PERP"
```

Failed to reproduce this one on strings on OpenJDK 17, including those with two-byte chars. It could be a JDK bug.

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**: 6.5
- **OS**:
- **Browser**:
```

### Additional context

_No response_

### LINKED ISSUE #2491: AsyncFilteredRecordCursorFactory concurrency issue
### Describe the bug

Following error appears on the log when running the below reproducer:
```
2022-08-30T04:12:13.410121Z I i.q.g.SqlCompiler plan [q=`select-choose column1 from (select [column1] from test1 timestamp (timestamp) where to_lowercase(column1) = '0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c')`, fd=98]
2022-08-30T04:12:13.410232Z I i.q.c.p.PGConnectionContext parse [fd=98, q=SELECT column1 FROM 'test1' WHERE to_lowercase(column1) = '0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c']
2022-08-30T04:12:13.410433Z C i.q.g.e.t.AsyncFilteredRecordCursor unexpected error [ex=
java.lang.StringIndexOutOfBoundsException: index 22, length 22
	at java.lang.String.checkIndex(String.java:4563)
	at java.lang.AbstractStringBuilder.charAt(AbstractStringBuilder.java:351)
	at java.lang.StringBuilder.charAt(StringBuilder.java:91)
	at io.questdb.std.str.StringSink.charAt(StringSink.java:71)
	at io.questdb.std.Chars.equalsChars(Chars.java:767)
	at io.questdb.std.Chars.equals(Chars.java:193)
	at io.questdb.std.Chars.equalsNc(Chars.java:323)
	at io.questdb.griffin.engine.functions.eq.EqStrFunctionFactory$ConstCheckFunc.getBool(EqStrFunctionFactory.java:115)
	at io.questdb.griffin.engine.table.AsyncFilteredRecordCursorFactory.filter(AsyncFilteredRecordCursorFactory.java:168)
	at io.questdb.cairo.sql.async.PageFrameReduceJob.reduce(PageFrameReduceJob.java:175)
	at io.questdb.cairo.sql.async.PageFrameReduceJob.consumeQueue(PageFrameReduceJob.java:132)
	at io.questdb.cairo.sql.async.PageFrameReduceJob.consumeQueue(PageFrameReduceJob.java:106)
	at io.questdb.cairo.sql.async.PageFrameSequence.stealWork(PageFrameSequence.java:391)
	at io.questdb.cairo.sql.async.PageFrameSequence.dispatch(PageFrameSequence.java:370)
	at io.questdb.cairo.sql.async.PageFrameSequence.next(PageFrameSequence.java:295)
	at io.questdb.griffin.engine.table.AsyncFilteredRecordCursor.fetchNextFrame(AsyncFilteredRecordCursor.java:189)
	at io.questdb.griffin.engine.table.AsyncFilteredRecordCursor.hasNext(AsyncFilteredRecordCursor.java:108)
	at io.questdb.cutlass.pgwire.PGConnectionContext.sendCursor0(PGConnectionContext.java:2485)
	at io.questdb.cutlass.pgwire.PGConnectionContext.sendCursor(PGConnectionContext.java:2480)
	at io.questdb.cutlass.pgwire.PGConnectionContext$PGConnectionBatchCallback.postCompile(PGConnectionContext.java:2638)
	at io.questdb.griffin.SqlCompiler.compileBatch(SqlCompiler.java:941)
	at io.questdb.cutlass.pgwire.PGConnectionContext.processQuery(PGConnectionContext.java:2288)
	at io.questdb.cutlass.pgwire.PGConnectionContext.parse(PGConnectionContext.java:1562)
	at io.questdb.cutlass.pgwire.PGConnectionContext.handleClientOperation(PGConnectionContext.java:418)
	at io.questdb.cutlass.pgwire.PGJobContext.handleClientOperation(PGJobContext.java:81)
	at io.questdb.cutlass.pgwire.PGWireServer$1.lambda$$0(PGWireServer.java:83)
	at io.questdb.network.AbstractIODispatcher.processIOQueue(AbstractIODispatcher.java:166)
	at io.questdb.cutlass.pgwire.PGWireServer$1.run(PGWireServer.java:110)
	at io.questdb.mp.Worker.run(Worker.java:118)
]
```

### To reproduce

```
package test

import (
	"fmt"
	"sync"
	"sync/atomic"
	"testing"

	"gorm.io/driver/postgres"
	"gorm.io/gorm"
)

func TestInit(t *testing.T) {
	conStr := fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=%s", "localhost", "8812", "admin", "quest", "qdb", "disable")
	db, err := gorm.Open(postgres.Open(conStr), &gorm.Config{})
	if err != nil {
		t.Fatal(err)
	}

	err = db.Exec(`DROP TABLE IF EXISTS 'test1';`).Error
	if err != nil {
		t.Fatal(err)
	}

	err = db.Exec(`	CREATE TABLE 'test1' (
		column1 SYMBOL capacity 256 CACHE index capacity 256,
		timestamp TIMESTAMP
	) timestamp (timestamp) PARTITION BY HOUR;`).Error
	if err != nil {
		t.Fatal(err)
	}

	for i := 1; i < 2000; i++ {
		insertQuery := fmt.Sprintf(`INSERT INTO test1 (column1,timestamp) VALUES ('0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c','2022-08-28T06:25:%02dZ');`, i%60)
		err := db.Exec(insertQuery).Error
		if err != nil {
			fmt.Println(i, err)
		}
	}
	fmt.Println("DONE")
}

func TestSelect(t *testing.T) {
	conStr := fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=%s", "localhost", "8812", "admin", "quest", "qdb", "disable")
	db, err := gorm.Open(postgres.Open(conStr), &gorm.Config{})
	if err != nil {
		t.Fatal(err)
	}

	selectQuery := `SELECT column1 FROM 'test1' WHERE to_lowercase(column1) = '0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c'`
	var wg sync.WaitGroup
	var c int32 = 0
	for i := 0; i < 100; i++ {
		wg.Add(1)
		go func(i int) {
			err := db.Exec(selectQuery).Error
			if err != nil {
				fmt.Println(i, err)
				atomic.AddInt32(&c, 1)
			}
			wg.Done()
		}(i)
	}
	wg.Wait()
	fmt.Println("DONE", "fail:", c)
}
```

### Expected Behavior

No error, each query should get a response.

### Environment

```markdown
- **QuestDB version**:
- **OS**:
- **Browser**:
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by glasstiger**
Fixes #2491
Fixes #2454

### CODE REVIEW

**Comment by questdb-butler** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Please update the PR title to match this format:
`type(subType): description`

Where `type` is one of:
`feat`, `fix`, `chore`, `docs`, `style`, `refactor`, `perf`, `test`, `build`, `ci`, `chore`, `revert`

And: `subType` is one of:
`build`, `sql`, `log`, `mig`, `core`, `ilp`, `pgwire`, `http`, `conf`

For Example:

```
perf(sql): improve pattern matching performance for SELECT sub-queries
```
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against fe91b0fb3882a2b00697a968c5d46b3b46634942
</p>

**Comment by puzpuzpuz** [comment]
@glasstiger this bug appears to be a silly race in async offload's work stealing logic: multiple threads having `workerId=-1` may be reducing different tasks within the same frame sequence (thus, the same filter `Function`) simultaneously. Here is a quick & dirty attempt to fix the bug:
https://github.com/puzpuzpuz/questdb/commit/1b3e562987ff4a18d8006566b33edc7a8439ed86#diff-8f4ab95dda2f482d6c717fc3474211ccbd4803bec790f43d04ea08a722735404R132-R141

Feel free to ignore this branch and take a different approach to fix this race.

File: core/src/main/java/io/questdb/griffin/engine/table/AsyncFilteredRecordCursorFactory.java [hunk 1]
```diff
@@ -156,11 +157,22 @@ public boolean hasDescendingOrder() {
...
+            PageAddressCacheRecord record,
+            PageFrameReduceTask task,
+            @Nullable PageFrameSequence<?> stealingFrameSequence
+    ) {
         final DirectLongList rows = task.getRows();
         final long frameRowCount = task.getFrameRowCount();
         final FilterAtom atom = task.getFrameSequence(FilterAtom.class).getAtom();
-        final Function filter = atom.getFilter(workerId);
+        final Function filter;
+        if (stealingFrameSequence != null && stealingFrameSequence != task.getFrameSequence(FilterAtom.class)) {
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/engine/table/AsyncFilteredRecordCursorFactory.java:170]: What would happen in the following scenario?

A user has configured a dedicated PG Wire pool. Then they start sending queries with a filter both over PG Wire and REST API. Those queries get executed on threads with 0 id on both PG Wire and shared worker pool and then both of them start work stealing. They would end with accessing the same filter concurrently.

Potentially, to protect from that we'd need to make sure that all worker ids are unique. In this case, we could use a single array of filter functions.

File: core/src/main/java/io/questdb/cairo/sql/async/PageFrameReduceJob.java [hunk 2]
```diff
@@ -86,6 +87,10 @@ public PageFrameReduceJob(
         }
     }
 
+    private static int getCallerThreadId() {
+        final Thread th = Thread.currentThread();
+        return th instanceof FixedThread ? ((FixedThread)th).getWorkerId() : -1;
+    }
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cairo/sql/async/PageFrameReduceJob.java:93]: nit: missing newline after this line.

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 100 / 114 (87.72%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/sql/async/PageFrameSequence.java|7|12|58.33%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|24|31|77.42%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilterAtom.java|37|39|94.87%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/sql/async/PageFrameReduceJob.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursorFactory.java|10|10|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 100 / 114 (87.72%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/sql/async/PageFrameSequence.java|7|12|58.33%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|24|31|77.42%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilterAtom.java|37|39|94.87%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/sql/async/PageFrameReduceJob.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursorFactory.java|10|10|100.00%|

**Pull request merged**